File: tutorials/src/tutorial7.scala

import java.util.Collection

import javafx.print.Collation


abstract class Set[A]
{
  var set: List[A]
  def addElement(newV : A) :  Boolean
  def removeElement(toRemove : A): Any
  def isEmpty(): Boolean
  def setSize(): Int
  def clearAll(): Unit
  def hashCode():Int
  def iterator(): Iterator[A]
  def contains(toFind : A): Boolean
  def containsAll(elem: Collection[A]):Boolean
  def toArray():Array[A]

  def diff(s1: Set[], s2 : Set[A]): Set[A]
  def sum(s1: Set[A], s2 : Set[A]): Set[A]
  def intersec(s1: Set[A], s2 : Set[A]): Set[A]


}




abstract class Map
{
  type Value
  type Key
  var value: List[Value] = List()
  var key: List[Key] = List()

  def addElement(k: Key, v: Value): Boolean
  def getValue(k:Key):Value
  def mapSize(): Int
  def isEmpty(): Boolean
  def clearAll(): Unit
  def removeElement(toRemove: Key): Value
  def hashCode():Int
  def newValue(k: Key, newV : Value): Value
  def values():Collection[Value]
  def keySet():Set[Key]
  def containsKey(k:Key):Boolean
  def containsValue(v:Value):Boolean
}

//TASK3

// LinkedList uses a doubly linked list to store elements. Can act both like list and queue because
// of its implementation.Each element is called node and it has two parts-one with data, and one with adress.
//They are connected one by one with a pointer,pointing to next one.That fact makes adding and removing
//elements fast, but access is more complicated, because we have to iterate through it.

//ArrayList uses dynamic array.It implements List only. Access to elements is easy because of indexing (direct access).
//While adding and removing elements , all the bits are shifted, because of internal use of array.

File: tutorials/src/Tutorial7.java
import java.util.ArrayList;
import java.util.LinkedList;

public class Tutorial7 {
    public static void main(String args[])
    {
        //ArrayList internally uses a dynamic array to store the elements, where we can apply a lot of modifications like
        //resizing it automatically,adding and removing items in dynamic way. LinkedList is a linear structure
        //where each element(called node) is an object with adress part and data part, and they are connected(linked) to each other
        //one by one by pointers and adresses(it uses DoublyLinkedList to store those elements).It implements both List
        //and Queque and can be used as both, while ArrayList implements just List.
        //In linkedlist its easier to insert and delete the elements, because unlikely to the ArrayList
        // we dont have to 'shift' other elements after/before such an operation, thats why the manipulation are faster then in ArrayList.
        //But on the other hand, ArrayList is faster when it comes to storing and accessing data since we can
        //access them directly because of index based system, but with LinkedList we have to travers through all elements to find one.
        //Simple implementation of both:

        LinkedList<String>  ourList = new LinkedList<String>();

        ourList.add("Anna");
        ourList.add("Barbra");
        ourList.add("Camilla");
        ourList.add("Daniela");

        System.out.println("Linked List before removing"+ourList);

        ourList.remove("Barbra");
        ourList.removeFirst();
        ourList.removeLast();

        System.out.println("Linked list after removing: " + ourList);
        ArrayList<Integer> arrlist = new ArrayList<Integer>();

        for (int i = 1; i <= 5; i++){
            arrlist.add(i);
        }
        arrlist.add(16);
        System.out.println("Array List before removing"+arrlist);
        arrlist.remove(4);
        System.out.println("Array list after removing: " +arrlist);
    }

}
